net: wifi: shell: add enterprise support for station

Rex-Chen-NXP opened the pull request:
Add EAP-TLS, EAP-PEAP-MSCHAPv2, EAP-PEAP-GTC, EAP-TTLS-MSCHAPv2, EAP-PEAP-TLS, EAP-TLS-SHA256  enterprise wpa2 and wpa3 suiteb support for station.

Rex-Chen-NXP committed:
net: wifi: shell: add enterprise support for station

Add EAP-TLS, EAP-PEAP-MSCHAPv2, EAP-PEAP-GTC, EAP-TTLS-MSCHAPv2,
EAP-PEAP-TLS, EAP-TLS-SHA256  enterprise wpa2 and wpa3 suiteb
support for station.

Signed-off-by: Rex Chen <rex.chen_1@nxp.com>

Rex-Chen-NXP commented:
Hi @jukkar  @krish2718 
Could you help to review, thanks!

jukkar commented on modules/hostap/src/supp_api.c:
```diff
@@ -430,6 +430,105 @@ static int wpas_config_process_blob(struct wpa_config *config, char *name, uint8
 }
+
+static const struct wifi_cipher_desc ciphers[] = {
+	{WPA_CAPA_ENC_CCMP_256, "CCMP-256"}, {WPA_CAPA_ENC_GCMP_256, "GCMP-256"},
+	{WPA_CAPA_ENC_CCMP, "CCMP"},         {WPA_CAPA_ENC_GCMP, "GCMP"},
+	{WPA_CAPA_ENC_TKIP, "TKIP"},         {WPA_CAPA_ENC_WEP104, "WEP104"},
+	{WPA_CAPA_ENC_WEP40, "WEP40"}};
```
These look bad, could you align them and put one line in a row. So
```
	{WPA_CAPA_ENC_CCMP_256, "CCMP-256"},
	{WPA_CAPA_ENC_GCMP_256, "GCMP-256"},
	{WPA_CAPA_ENC_CCMP, "CCMP"},
	{WPA_CAPA_ENC_GCMP, "GCMP"},
	{WPA_CAPA_ENC_TKIP, "TKIP"},
	{WPA_CAPA_ENC_WEP104, "WEP104"},
	{WPA_CAPA_ENC_WEP40, "WEP40"},
};
```

Rex-Chen-NXP replied:
Updated as comment.

jukkar commented on modules/hostap/src/supp_api.c:
```diff
@@ -443,6 +542,14 @@ static int wpas_add_and_config_network(struct wpa_supplicant *wpa_s,
+	struct wifi_eap_cipher_config cipher_config = {NULL,   "DEFAULT:!EXP:!LOW", "CCMP",
+						       "CCMP", "AES-128-CMAC",      NULL};
+	char *method;
+	char phase1[256] = {0};
```
Do we need 256 chars for this, just wanted to confirm because this seems quite long string.

Rex-Chen-NXP replied:
Yes, indeed, 256 bytes consume too many stack space, I checked supplicant code, 128 bytes to store phase1 info should enough.

jukkar commented on modules/hostap/src/supp_api.c:
```diff
@@ -611,11 +756,44 @@ static int wpas_add_and_config_network(struct wpa_supplicant *wpa_s,
+			if (params->security == WIFI_SECURITY_TYPE_EAP_PEAP_MSCHAPV2 ||
+			    params->security == WIFI_SECURITY_TYPE_EAP_PEAP_GTC ||
+			    params->security == WIFI_SECURITY_TYPE_EAP_PEAP_TLS) {
+				snprintf(phase1, sizeof(phase1),
```
I would recommend using `snprintk` here as that should have smaller memory consumption

Rex-Chen-NXP replied:
Updated as comment.

jukkar commented on samples/net/wifi/test_certs/ca2.pem:
```diff
@@ -0,0 +1,29 @@
+-----BEGIN CERTIFICATE-----
```
Could you place the certificate changes to a separate commit, also document in the readme file how they were created, and also name the files better, perhaps adding `station` string to name or similar.

Rex-Chen-NXP replied:
Yes, of course, need to add more detail info for new added files and raise a separate commit. This will be changed after the details of phase2 discussion.
Hi @krish2718 , This phase2 certificate files same as phase1, as it only for test, I just copied it. should we generate a new set of certificate files?

krish2718 replied:
> Hi @krish2718 , This phase2 certificate files same as phase1, as it only for test, I just copied it. should we generate a new set of certificate files?

Yes, please generate separate certs.

Rex-Chen-NXP replied:
Updated all certs.

jukkar commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -490,19 +508,22 @@ static int __wifi_args_to_params(const struct shell *sh, size_t argc, char *argv
+						     {"timeout", required_argument, 0, 't'},
+						     {"anon-id", required_argument, 0, 'a'},
+						     {"key-passwd", required_argument, 0, 'K'},
+						     {"suiteb-type", required_argument, 0, 'S'},
+						     {"eap-version", required_argument, 0, 'V'},
+						     {"eap-identity", required_argument, 0, 'I'},
+						     {"eap-password", required_argument, 0, 'P'},
```
Could you add these to the option array without changing the alignment. These are now quite far right which is not really needed.

Rex-Chen-NXP replied:
Updated as comments.

Rex-Chen-NXP commented:
Hi @jukkar @krish2718 

For EAP-PEAP-TLS, if the keys(client-key or server-key) for pharse1 and pharse2 is encrypted, we need to configure key passwords for pharse1 and pharse2. 
How should we add usage for ‘wifi connect’ cmd to configure the key passwords? I prepare to use twice ‘-k’ to set it, cmd example show below:
wifi connect -s EapNet -c 149 -k 17 -w 2 -a client1 -K key_passwrd_1 -K key_passwrd_2  -I user -P password

Could you provide some suggestions?

krish2718 commented:
> Hi @jukkar @krish2718
> 
> For EAP-PEAP-TLS, if the keys(client-key or server-key) for pharse1 and pharse2 is encrypted, we need to configure key passwords for pharse1 and pharse2. How should we add usage for ‘wifi connect’ cmd to configure the key passwords? I prepare to use twice ‘-k’ to set it, cmd example show below: wifi connect -s EapNet -c 149 -k 17 -w 2 -a client1 -K key_passwrd_1 -K key_passwrd_2 -I user -P password
> 
> Could you provide some suggestions?

Wouldn't that be confusing relying on the order of args to map to phase1/2? May be cleanly add `-K1/-K2` ? `-K` still backwards compatible to `-K1`?

Rex-Chen-NXP commented:
> > Hi @jukkar @krish2718
> > For EAP-PEAP-TLS, if the keys(client-key or server-key) for pharse1 and pharse2 is encrypted, we need to configure key passwords for pharse1 and pharse2. How should we add usage for ‘wifi connect’ cmd to configure the key passwords? I prepare to use twice ‘-k’ to set it, cmd example show below: wifi connect -s EapNet -c 149 -k 17 -w 2 -a client1 -K key_passwrd_1 -K key_passwrd_2 -I user -P password
> > Could you provide some suggestions?
> 
> Wouldn't that be confusing relying on the order of args to map to phase1/2? May be cleanly add `-K1/-K2` ? `-K` still backwards compatible to `-K1`?

Yes, of course, will try.
In order to implement with a more better solution. I want to insert discussion about sap enterprise mode here, because also face same situation for sap enterprise mode. 
we need to set users identity and passwords for multi users, thus can filter out target users, so also need to set user identity and password multi times, is there other suggestion if take this into consideration?

Rex-Chen-NXP commented on include/zephyr/net/wifi.h:
```diff
@@ -76,6 +86,86 @@ enum wifi_security_type {
 	/** @endcond */
 };
 
+enum wifi_eap_type {
```
Yes, added documentation for all enum items

Rex-Chen-NXP commented:
> > > Hi @jukkar @krish2718
> > > For EAP-PEAP-TLS, if the keys(client-key or server-key) for pharse1 and pharse2 is encrypted, we need to configure key passwords for pharse1 and pharse2. How should we add usage for ‘wifi connect’ cmd to configure the key passwords? I prepare to use twice ‘-k’ to set it, cmd example show below: wifi connect -s EapNet -c 149 -k 17 -w 2 -a client1 -K key_passwrd_1 -K key_passwrd_2 -I user -P password
> > > Could you provide some suggestions?
> > 
> > 
> > Wouldn't that be confusing relying on the order of args to map to phase1/2? May be cleanly add `-K1/-K2` ? `-K` still backwards compatible to `-K1`?
> 
> Yes, of course, will try. In order to implement with a more better solution. I want to insert discussion about sap enterprise mode here, because also face same situation for sap enterprise mode. we need to set users identity and passwords for multi users, thus can filter out target users, so also need to set user identity and password multi times, is there other suggestion if take this into consideration?

Hi @krish2718 
Could you add comments about the above discussed shell usage for sap users identity and passwords, also key passwords for station case?

krish2718 commented:
> > > > Hi @jukkar @krish2718
> > > > For EAP-PEAP-TLS, if the keys(client-key or server-key) for pharse1 and pharse2 is encrypted, we need to configure key passwords for pharse1 and pharse2. How should we add usage for ‘wifi connect’ cmd to configure the key passwords? I prepare to use twice ‘-k’ to set it, cmd example show below: wifi connect -s EapNet -c 149 -k 17 -w 2 -a client1 -K key_passwrd_1 -K key_passwrd_2 -I user -P password
> > > > Could you provide some suggestions?
> > > 
> > > 
> > > Wouldn't that be confusing relying on the order of args to map to phase1/2? May be cleanly add `-K1/-K2` ? `-K` still backwards compatible to `-K1`?
> > 
> > 
> > Yes, of course, will try. In order to implement with a more better solution. I want to insert discussion about sap enterprise mode here, because also face same situation for sap enterprise mode. we need to set users identity and passwords for multi users, thus can filter out target users, so also need to set user identity and password multi times, is there other suggestion if take this into consideration?
> 
> Hi @krish2718 Could you add comments about the above discussed shell usage for sap users identity and passwords, also key passwords for station case?
> 
> A solution for this from my side. Use the long option feature of shell to achieve it, all the long opt strings of one kind of opt mapping to same value, for example , both key-passwd1 and key-passwd2 mapping to 'k', then we can save key-passwd1 and key-passwd2 info base on opt 'k', and for user, shell usage is clearly.
> 
> long options parameter set as below: static const struct option long_options[] = { {"key-passwd1", required_argument, 0, 'K'}, {"key-passwd2", required_argument, 0, 'K'},
> 
> ```
> {"eap-identity1", required_argument, 0, 'I'},
> {"eap-identity2", required_argument, 0, 'I'},
> ...
> {"eap-identity8", required_argument, 0, 'I'},
> 
> {"eap-password1", required_argument, 0, 'P'},
> {"eap-password2", required_argument, 0, 'P'},
> ...
> {"eap-password8", required_argument, 0, 'P'},
> ```
> 
> };
> 
> Shell usage: wifi connect --key-passwd1 key1 --key-passwd2 key2 wifi ap enable --eap-identity1 id1... --eap-identity8 id8 --eap-password1 passwd1 ... --eap-password8 passwd8
> 
> If you approve this solution, I will try.

LGTM, I don't see how else we can achieve this, Shell usage is primarily for testing (both users and developers), so, this is fine.

Rex-Chen-NXP commented:
Hi @krish2718 
The  private key is encrypt, could you share me the private key password of the current private key, or which doc recorded the related information, then I can verify base on it.

Rex-Chen-NXP commented:
Hi @krish2718  
Could you pls also review the PR? Thanks!

krish2718 commented:
> Hi @krish2718 The private key is encrypt, could you share me the private key password of the current private key, or which doc recorded the related information, then I can verify base on it.

https://docs.zephyrproject.org/latest/connectivity/networking/api/wifi.html#wi-fi-enterprise-test-x-509-certificate-header-generation

krish2718 commented on include/zephyr/net/wifi.h:
```diff
@@ -68,6 +68,16 @@ enum wifi_security_type {
+	WIFI_SECURITY_TYPE_EAP_PEAP_GTC,
+	/** EAP TTLS MSCHAPV2 security - Enterprise. */
+	WIFI_SECURITY_TYPE_EAP_TTLS_MSCHAPV2,
+	/** EAP PEAP TLS security - Enterprise. */
+	WIFI_SECURITY_TYPE_EAP_PEAP_TLS,
```
EAP-PEAP should be fine, we can drop the TLS

Rex-Chen-NXP replied:
Updated as comment

krish2718 commented on include/zephyr/net/wifi.h:
```diff
@@ -76,6 +86,93 @@ enum wifi_security_type {
+
+/** @brief group mgmt cipher types. */
+enum wifi_group_mgmt_cipher_type {
+	/* 128-bit Broadcast/Multicast Integrity Protocol
+	 * Cipher-based Message Authentication Code .
+	 */
+	WPA_CAPA_ENC_BIP,
+	/* 128-bit Broadcast/Multicast Integrity Protocol
+	 * Galois Message Authentication Code .
+	 */
+	WPA_CAPA_ENC_BIP_GMAC_128,
+	/* 256-bit Broadcast/Multicast Integrity Protocol
+	 * Galois Message Authentication Code .
+	 */
```
```suggestion
	/** 128-bit Broadcast/Multicast Integrity Protocol
	 * Cipher-based Message Authentication Code .
	 */
	WPA_CAPA_ENC_BIP,
	/** 128-bit Broadcast/Multicast Integrity Protocol
	 * Galois Message Authentication Code .
	 */
	WPA_CAPA_ENC_BIP_GMAC_128,
	/** 256-bit Broadcast/Multicast Integrity Protocol
	 * Galois Message Authentication Code .
	 */
```

Rex-Chen-NXP replied:
Updated as comment

Rex-Chen-NXP committed:
net: wifi: shell: update enterprise mode certificate files

Update enterprise mode certificate files to support phase1 and
phase2.

Signed-off-by: Rex Chen <rex.chen_1@nxp.com>

Rex-Chen-NXP commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -500,7 +518,26 @@ static int __wifi_args_to_params(const struct shell *sh, size_t argc, char *argv
 		{"timeout", required_argument, 0, 't'},
 		{"anon-id", required_argument, 0, 'a'},
-		{"key-passwd", required_argument, 0, 'K'},
+		{"key1-pwd", required_argument, 0, 'K'},
```
Add long opt to support multi key passwords , also for user identities\passwords.

Rex-Chen-NXP commented:
Hi @krish2718  @jukkar  and all reviewers
The changes of the update:
A. Add input items for key pwds of prv key, user ids and pwds.
B. Enhance EAP phase2 part code.

Rex-Chen-NXP commented:
Hi @krish2718 
Let's call CA1\CERT1\KEY1 for phase1, and CA2\CERT2\KEY2 for phase2.
I updated cert files. The reason why I make change:
A. Need support EAP phase2, it should be exist 2 set of cert files(CA\CERT\KEY), but hostap side NO veriable to maintain CA2, I think the CERT1\KEY1 and CERT2\KEY2 should both generated base on the SAME CA\CA-KEY,
So the 2 set of cert files should generated by same ca.
B. I tried to copy to make CA2\CERT2\KEY2 same as CA1\CERT1\KEY1, thus can reach the above point1, but unfortunately, the cert files from you upstreamed previouly can't works ok for suite-B and suite-B 192, failed reason is 'Insufficient RSA modulus size'.
Base on the above 2 points, I updated the cert files which priv key is open mode.

krish2718 commented:
> Hi @krish2718 Let's call CA1\CERT1\KEY1 for phase1, and CA2\CERT2\KEY2 for phase2, and my local code has supported sap and station enterprise mode, so both sap and station base on zephyr platform. I updated cert files. The reason why I make change: A. Need support EAP phase2, it should be exist 2 set of cert files(CA\CERT\KEY), but hostap side NO veriable to maintain CA2, I think the CERT1\KEY1 and CERT2\KEY2 should both generated base on the SAME CA\CA-KEY, So the 2 set of cert files should generated by same ca. B. I tried to copy to make CA2\CERT2\KEY2 same as CA1\CERT1\KEY1, thus can reach the above point1, but unfortunately, the cert files from you upstreamed previouly can't works ok for suite-B and suite-B 192, failed reason is 'Insufficient RSA modulus size'. Base on the above 2 points, I updated the cert files which priv key is open mode.

IIUC, the private keys updated in this PR doesn't have any passphrase. Is that what you mean by `open mode`?

krish2718 commented on include/zephyr/net/wifi_mgmt.h:
```diff
@@ -462,6 +462,22 @@ struct wifi_connect_req_params {
 	const uint8_t *key_passwd;
 	/** Private key passwd length */
 	uint8_t key_passwd_length; /* Max 128 */
+	/** private key2 passwd */
+	const uint8_t *key2_passwd;
+	/** key2 passwd length */
+	uint8_t key2_passwd_length; /* Max 128 */
+	/** suiteb or suiteb-192 */
+	uint8_t suiteb_type;
+	/** eap version */
+	uint8_t eap_ver;
+	/** Identity for EAP */
+	const uint8_t *eap_identity;
+	/** eap identity length */
+	uint8_t eap_id_length; /* Max 64 */
+	/** Password string for EAP. */
+	const uint8_t *eap_password;
+	/** eap passwd length */
+	uint8_t eap_passwd_length; /* Max 128 */
```
all the length comments can be clubbed to the main comment for them to come up in doxygen.

Rex-Chen-NXP replied:
Updated as comment.

krish2718 commented on modules/hostap/src/supp_api.c:
```diff
@@ -430,8 +430,120 @@ static int wpas_config_process_blob(struct wpa_config *config, char *name, uint8
+	return -EINVAL;
+}
+
+static int is_eap_valid_security(int security)
+{
+	if (security == WIFI_SECURITY_TYPE_EAP_TLS ||
+	    security == WIFI_SECURITY_TYPE_EAP_PEAP_MSCHAPV2 ||
+	    security == WIFI_SECURITY_TYPE_EAP_PEAP_GTC ||
+	    security == WIFI_SECURITY_TYPE_EAP_TTLS_MSCHAPV2 ||
+	    security == WIFI_SECURITY_TYPE_EAP_PEAP_TLS ||
+	    security == WIFI_SECURITY_TYPE_EAP_TLS_SHA256) {
+		return 1;
+	}
+
+	return 0;
+}
```
```suggestion
static int is_eap_valid_security(int security)
{
	return (security == WIFI_SECURITY_TYPE_EAP_TLS ||
			security == WIFI_SECURITY_TYPE_EAP_PEAP_MSCHAPV2 ||
			security == WIFI_SECURITY_TYPE_EAP_PEAP_GTC ||
			security == WIFI_SECURITY_TYPE_EAP_TTLS_MSCHAPV2 ||
			security == WIFI_SECURITY_TYPE_EAP_PEAP_TLS ||
			security == WIFI_SECURITY_TYPE_EAP_TLS_SHA256);
}
```

Rex-Chen-NXP replied:
Updated as comment.

krish2718 commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -2807,16 +2895,22 @@ SHELL_STATIC_SUBCMD_SET_CREATE(wifi_commands,
 		  "[-t, --timeout]: Timeout for the connection attempt (in seconds).\n"
 		  "[-a, --anon-id]: Anonymous identity for enterprise mode.\n"
-		  "[-K, --key-passwd]: Private key passwd for enterprise mode.\n"
+		  "[-K, --key1-pwd for eap phase1 or --key2-pwd for eap phase2]:\n"
+		  "Private key passwd for enterprise mode.\n"
+		  "[-S, --suiteb-type]: 1:suiteb, 2:suiteb-192.\n"
+		  "[-V, --eap-version]: 0 or 1.\n"
+		  "[-I, --eap-id1 ... eap-id8]: Client Identity.\n"
+		  "[-P, --eap-pwd1 ... eap-pwd8]: Client Password.\n"
```
as these are optional, please mention defaults as well.

krish2718 reviewed (commented):
can you also update the docs `wifi.rst`? May be add separate sections single phase and phase1 and phase2 separate certs?

Rex-Chen-NXP commented:
> > Hi @krish2718 Let's call CA1\CERT1\KEY1 for phase1, and CA2\CERT2\KEY2 for phase2, and my local code has supported sap and station enterprise mode, so both sap and station base on zephyr platform. I updated cert files. The reason why I make change: A. Need support EAP phase2, it should be exist 2 set of cert files(CA\CERT\KEY), but hostap side NO veriable to maintain CA2, I think the CERT1\KEY1 and CERT2\KEY2 should both generated base on the SAME CA\CA-KEY, So the 2 set of cert files should generated by same ca. B. I tried to copy to make CA2\CERT2\KEY2 same as CA1\CERT1\KEY1, thus can reach the above point1, but unfortunately, the cert files from you upstreamed previouly can't works ok for suite-B and suite-B 192, failed reason is 'Insufficient RSA modulus size'. Base on the above 2 points, I updated the cert files which priv key is open mode.
> 
> IIUC, the private keys updated in this PR doesn't have any passphrase. Is that what you mean by `open mode`?

Yes, it is

Rex-Chen-NXP commented:
> can you also update the docs `wifi.rst`? May be add separate sections single phase and phase1 and phase2 separate certs?

A. can you also update the docs `wifi.rst`?
----Yes, should add detail info of added phase2 certs.
B. May be add separate sections single phase and phase1 and phase2 separate certs?
---Actually previous I also want to give option for customer to select single phase mode or both phase1 and phase2 mode , but late I think  single phase is a subset case of both phase1 and phase2, if we support both phase1 and phase2, customer only need to generate certs and can make all security types work. We default support  both phase1 and phase2 mode after this PR merged. So I don't add a separate sections for phase2.

jukkar commented:
The pr looks ok, please fix the merge conflicts.

Rex-Chen-NXP commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -2807,16 +2895,23 @@ SHELL_STATIC_SUBCMD_SET_CREATE(wifi_commands,
+		  "Private key passwd for enterprise mode. Default no password for private key.\n"
+		  "[-S, --suiteb-type]: 1:suiteb, 2:suiteb-192. Default 0: not suiteb mode.\n"
+		  "[-V, --eap-version]: 0 or 1. Default 1: eap version 1.\n"
+		  "[-I, --eap-id1 ... eap-id8]: Client Identity. Default no eap identity.\n"
+		  "[-P, --eap-pwd1 ... eap-pwd8]: Client Password.\n"
```
Yes, indeed, actually I make mistake, 8 users is for sap enterprise, station case should only need 1

Rex-Chen-NXP commented on doc/connectivity/networking/api/wifi.rst:
```diff
@@ -38,21 +38,21 @@ module.
 .. code-block:: console
 
     uart:~$ wifi connect -s <SSID> -k 5 -a anon -K whatever
+    uart:~$ wifi connect -s <SSID> -c 149 -k 17 -w 2 -a client1 --key1-pwd whatever --key2-pwd whatever --eap-id1 id1 --eap-pwd1 pwd1
```
Okay, will move EAP-TLS

Rex-Chen-NXP commented:
Hi @krish2718  @jukkar 
Updated comments from you, and fix conflict, could you help approve and merge the PR?

Rex-Chen-NXP commented:
Hi @krish2718 @jukkar 
Could you help approve and merge PR, then I will upstream sap enterprise patch.
